font-variant-alternates list support and case-insensitivity

mdubet opened the pull request:
#### fd6f66d0417561831691ccd5ea165b4062f7f4f4
<pre>
font-variant-alternates list support and case-insensitivity
<a href="https://bugs.webkit.org/show_bug.cgi?id=246811">https://bugs.webkit.org/show_bug.cgi?id=246811</a>
rdar://101638181

<a href="https://www.w3.org/TR/css-fonts-4/#font-variant-alternates-prop">https://www.w3.org/TR/css-fonts-4/#font-variant-alternates-prop</a>

CSS &quot;character-variant&quot; and &quot;styleset&quot; support a list of
feature names in CSS. Before this patch, only a single feature name was supported.

This patch also contains a change to make font family name matching
case insensitive for &quot;@font-feature-values&quot; at-rule.

Reviewed by Darin Adler.

* LayoutTests/TestExpectations:

The alternates-order.html layout test contains 9 specific tests related to font variant.
This patch fixes the remaining 2 out of 9 which were failing:
  - one because of the case-sensitivity in font name matching.
  - another because of the missing support for list value.

* LayoutTests/platform/gtk/TestExpectations:
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::makeStringByJoining):
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontFeatureValuesRule):
(WebCore::CSSFontSelector::lookupFontFeatureValues):
* Source/WebCore/css/CSSFontSelector.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontVariantAlternates):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator&lt;&lt;):
(WebCore::computeFeatureSettingsFromVariants):
* Source/WebCore/platform/text/TextFlags.h:
* Tools/TestWebKitAPI/Tests/WTF/WTFString.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/256702@main">https://commits.webkit.org/256702@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f8527593d1ee9ef9164c4ec08cdc6792184a3a5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96442 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5696 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29511 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105985 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166335 "An unexpected error occured. Recent messages:Pull request contains relevant changes; Failed to print configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100425 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5875 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34443 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88821 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102710 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102113 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4381 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83045 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31368 "Found 1 new test failure: imported/w3c/web-platform-tests/fetch/stale-while-revalidate/stale-image.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86223 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88114 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74267 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87427 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40172 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/19605 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/82823 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37845 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20996 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28299 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/4639 "The change is no longer eligible for processing.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4121 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43558 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/85501 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2226 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/929 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40266 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19271 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -78,6 +78,17 @@ UChar32 String::characterStartingAt(unsigned i) const
+    for (const auto& string : strings) {
+        if (!builder.isEmpty())
+            builder.append(separator);
+        builder.append(string);
```
This can be done more efficiently:
```
    for (auto& string : strings) {
        if (builder.isEmpty())
            builder.append(string);
        else
            builder.append(separator, string);
    }
```
Appending both in a single call results in less `StringBuilder` resizing.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -474,6 +474,8 @@ ALWAYS_INLINE String WARN_UNUSED_RETURN makeStringByReplacingAll(const String& s
 
 WTF_EXPORT_PRIVATE String WARN_UNUSED_RETURN makeStringByRemoving(const String&, unsigned position, unsigned lengthToRemove);
 
+WTF_EXPORT_PRIVATE String makeStringByJoining(const Vector<String>& strings, UChar separator);
```
I suggest making this function take a `Span<const String>` so it can be more widely used on collections other than `Vector`, such as `std::array` or `FixedVector`.

darinadler commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -251,11 +251,13 @@ void CSSFontSelector::addFontFeatureValuesRule(const StyleRuleFontFeatureValues&
     for (const auto& fontFamily : fontFeatureValuesRule.fontFamilies()) {
-        auto exist = m_featureValues.get(fontFamily);
+        // https://www.w3.org/TR/css-fonts-3/#font-family-casing
+        auto lowercased = fontFamily.string().convertToLowercaseWithoutLocale();
```
This fix is not mentioned in the bug title, but should be.

darinadler commented on Source/WebCore/css/CSSFontSelector.h:
```diff
@@ -138,7 +138,7 @@ class CSSFontSelector final : public FontSelector, public CSSFontFace::Client, p
     };
     HashMap<std::pair<AtomString, AtomString>, FontPaletteValues, PaletteMapHash> m_paletteMap;
-    HashMap<AtomString, Ref<FontFeatureValues>> m_featureValues;
+    HashMap<String, Ref<FontFeatureValues>> m_featureValues;
```
Might still be useful to use `AtomString` to save memory, even if it requires putting atom strings into the table.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -816,7 +816,10 @@ static RefPtr<CSSValue> consumeFontVariantAlternates(CSSParserTokenRange& range)
     auto parseSomethingWithoutError = [&range, &result]() {
         bool hasParsedSomething = false;
-        auto parseAndSetArgument = [&range, &hasParsedSomething] (auto& value) {
+        auto parseAndSetArgument = [&range, &hasParsedSomething] (std::optional<String>& value) {
```
Using a `std::optional<String>` is typically rare since `String` has a null value.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -77,10 +77,10 @@ WTF::TextStream& operator<<(TextStream& ts, FontVariantAlternates alternates)
+        if (!values.styleset.isEmpty())
+            append("styleset(", makeStringByJoining(values.styleset, ','), ")");
+        if (!values.characterVariant.isEmpty())
+            append("character-variant(", makeStringByJoining(values.characterVariant, ','), ")");
```
A little sad to have to build and discard an intermediate string here. We should seek idioms where we build the string up once, rather than building strings and then building strings from those strings.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -282,26 +282,29 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
+            };
+
+            // For styleset and character-variant, the tag name itself is the actual conveyor of information.
+            auto addFeatureTags = [&makeTag, &features, &lookupTags](const Vector<String>& names, const auto& tags, std::array<char, 2> codename) {
```
I think it’s probably better to just capture `[&]` rather than listing exactly what we want to capture.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -825,11 +828,26 @@ static RefPtr<CSSValue> consumeFontVariantAlternates(CSSParserTokenRange& range)
+            value = ident->stringValue();
+            return true;
+        };
+        auto parseListAndSetArgument = [&range, &hasParsedSomething] (Vector<String>& value) {
```
I think it’s better to use a return value rather than a separate bool and out argument.

We should consider `FixedVector<String>` because that will use less memory after initially creating it. `Vector<String>` is better if we might later append more.

mdubet committed:
font-variant-alternates list support and case-insensitivity
https://bugs.webkit.org/show_bug.cgi?id=246811
rdar://101638181

https://www.w3.org/TR/css-fonts-4/#font-variant-alternates-prop

CSS "character-variant" and "styleset" support a list of
feature names in CSS. Before this patch, only a single feature name was supported.

This patch also contains a change to make font family name matching
case insensitive for "@font-feature-values" at-rule.

Reviewed by Darin Adler.

* LayoutTests/TestExpectations:

The alternates-order.html layout test contains 9 specific tests related to font variant.
This patch fixes the remaining 2 out of 9 which were failing:
  - one because of the case-sensitivity in font name matching.
  - another because of the missing support for list value.

* LayoutTests/platform/gtk/TestExpectations:
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::makeStringByJoining):
* Source/WTF/wtf/text/WTFString.h:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontFeatureValuesRule):
(WebCore::CSSFontSelector::lookupFontFeatureValues):
* Source/WebCore/css/CSSFontSelector.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFontVariantAlternates):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator<<):
(WebCore::computeFeatureSettingsFromVariants):
* Source/WebCore/platform/text/TextFlags.h:
* Tools/TestWebKitAPI/Tests/WTF/WTFString.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/256702@main
